### PR TITLE
[FrontEnd] Loading frontends only when needed

### PR DESCRIPTION
**Pull request by nosovmik**
### Details:
 - On creation of FrontEndManager - search for available FrontEnd plugins but do not load them
 - Reworked logic for loading frontends using `FrontendManager::load_by_model`
     1) Check model extension. If it is '.xml' - try load with IR FE. If it is '.onnx' - try with ONNX FE. For others, go to next steps
     2) If 'input stream' is passed instead of file name - try IR, then ONNX, then others
     3) Check already loaded frontends: if there is suitable one ('supports' returns true) - use it
     4) Load unloaded frontends until suitable one is found
 - This significantly reduces 'ReadNetwork' time for most common case where IR is used
 - Testing is done only manually, no new auto-tests were created

### Tickets:
 - 61780, 68177, 69168

### CODE REVIEW

File: cmake/templates/ngraphConfig.cmake.in [hunk 1]
```diff
@@ -28,11 +28,11 @@
 #
 #   ngraph::frontend_manager               - nGraph frontend manager
 #
-#   ngraph_onnx_ngraph_frontend_FOUND      - True if the system has onnx_ngraph_frontend library
-#   ngraph::onnx_ngraph_frontend           - ONNX FrontEnd target (optional)
+#   ngraph_onnx_ov_frontend_FOUND          - True if the system has onnx_ov_frontend library
```
Comment (ilyachur) [inline cmake/templates/ngraphConfig.cmake.in:31]: I think for the backward compatibility we shouldn't rename variables. (it is obsolete cmake config)

File: model-optimizer/unit_tests/mock_mo_frontend/mock_mo_python_api/mock_mo_python_api.cpp [hunk 2]
```diff
@@ -5,7 +5,7 @@
 #include <pybind11/pybind11.h>
 #include <pybind11/stl.h>
 
-#include "../mock_mo_ngraph_frontend/mock_mo_frontend.hpp"
+#include "../mock_mo_ov_frontend/mock_mo_frontend.hpp"
```
Comment (ilyachur) [inline model-optimizer/unit_tests/mock_mo_frontend/mock_mo_python_api/mock_mo_python_api.cpp:8]: Can we avoid relative paths?

File: ngraph/frontend/frontend_manager/include/frontend_manager/frontend_manager.hpp [hunk 3]
```diff
@@ -27,10 +27,10 @@ class FRONTEND_API FrontEndManager final {
     FrontEndManager();
 
     /// \brief Default move constructor
-    FrontEndManager(FrontEndManager&&);
+    FrontEndManager(FrontEndManager&&) noexcept;
```
Comment (ilyachur) [inline ngraph/frontend/frontend_manager/include/frontend_manager/frontend_manager.hpp:30]: Why do you add `noexcept`?

Comment (nosovmik) [reply ngraph/frontend/frontend_manager/include/frontend_manager/frontend_manager.hpp]: It is not related to PR, just a 'IDE warning' fix, it is good style to declare move constructor/operator= as noexcept (some algorithms have more efficient implementations if such constructors are noexcept)

File: model-optimizer/unit_tests/mock_mo_frontend/mock_mo_python_api/mock_mo_python_api.cpp (hunk 2 above)
Comment (nosovmik) [reply model-optimizer/unit_tests/mock_mo_frontend/mock_mo_python_api/mock_mo_python_api.cpp]: Not relevant to this PR anymore

File: cmake/templates/ngraphConfig.cmake.in (hunk 1 above)
Comment (nosovmik) [reply cmake/templates/ngraphConfig.cmake.in]: Not applicable to this PR anymore

File: ngraph/frontend/frontend_manager/src/frontend_manager.cpp [hunk 4]
```diff
@@ -48,50 +69,140 @@ class FrontEndManager::Impl {
...
+        };
+        // Map between file extension and suitable frontend
+        static const std::map<std::string, PriorityFrontEndInfo> priority_frontends = {
+            {".xml", {"ir", std::string(FRONTEND_LIB_PREFIX) + "ir" + FRONTEND_LIB_SUFFIX}},
+            {".onnx", {"onnx", std::string(FRONTEND_LIB_PREFIX) + "onnx" + FRONTEND_LIB_SUFFIX}},
+        };
+        // List of prioritized frontends
+        std::list<PriorityFrontEndInfo> priority_list = {
+            {"ir", std::string(FRONTEND_LIB_PREFIX) + "ir" + FRONTEND_LIB_SUFFIX},
+            {"onnx", std::string(FRONTEND_LIB_PREFIX) + "onnx" + FRONTEND_LIB_SUFFIX}};
```
[the remark this answers is not in the document]
Comment (nosovmik) [reply ngraph/frontend/frontend_manager/src/frontend_manager.cpp]: Done, for ir, onnx, tf, paddle

File: ngraph/frontend/frontend_manager/src/frontend_manager.cpp [hunk 5]
```diff
@@ -17,27 +18,47 @@ using namespace ngraph::frontend;
 
 //----------- FrontEndManager ---------------------------
 class FrontEndManager::Impl {
+    std::mutex m_loading_mutex;
     std::vector<PluginHandle> m_loaded_libs;  // must be a first class member (destroyed last)
+    std::list<PluginDescription> m_not_loaded_plugins;
     std::map<std::string, FrontEndFactory> m_factories;
```
[the remark this answers is not in the document]
Comment (nosovmik) [reply ngraph/frontend/frontend_manager/src/frontend_manager.cpp]: PR has been reworked according to suggestion (priority is still handled outside)

File: ngraph/frontend/frontend_manager/src/frontend_manager.cpp [hunk 6]
```diff
@@ -17,27 +18,47 @@ using namespace ngraph::frontend;
...
     FrontEnd::Ptr load_by_framework(const std::string& framework) {
+        std::lock_guard<std::mutex> guard(m_loading_mutex);
+        // Load plugins until we found the right one
+        while (!m_factories.count(framework) && !m_not_loaded_plugins.empty()) {
+            auto desc = m_not_loaded_plugins.front();
+            auto loaded_data = desc.load();
+            m_loaded_libs.push_back(std::move(loaded_data.m_lib_handle));
+            m_factories.insert({loaded_data.m_plugin_info.m_name, loaded_data.m_plugin_info.m_creator});
+            m_not_loaded_plugins.pop_front();
+        }
```
[the remark this answers is not in the document]
Comment (nosovmik) [reply ngraph/frontend/frontend_manager/src/frontend_manager.cpp]: fixed

**Review by ilya-lavrenov — APPROVED** [review]

**Pull request merged**